### ISSUE TITLE
Update dependencies (1.53.0)

### ISSUE THREAD

**Issue by StefanKock** [hzi-braunschweig/SORMAS-Project#3584]
### Problem Description
Dependencies need updates at least to include bugfixes.

### Proposed Change

1. Check for easy version updates (after clearing up in #3449 with Payara modules).
2. Update maintainance versions
- hibernate-core: 5.4.18.Final -> 5.4.24.Final (contains performance improvements)
- ...
3. Document versions on hold in POM with reason.

### Additional Information

Last time done in #2459

**Comment by markusmann-vg** [comment]
The following list of dependencies with known vulnerabilities has been identified:
o jackson-databind:2.0.1 used by com.auth0-java-jwt-2.2.1.jar affected by 35 CVEs o com.fasterxml.jackson.core-jackson-databind-2.9.8.jar affected by 33 CVEs
o com.google.guava-guava-19.0.jar affected by CVE-2018-10237
o commons-beanutils-commons-beanutils-1.9.2.jar affected by CVE-2019-10086 o commons-beanutils.jar affected by CVE-2019-10086
o org.geotools-gt-shapefile-21.4.jar affected by CVE-2018-18749
o snakeyaml.jar affected by CVE-2017-18640
o vaadin-sass-compiler.jar affected by CVE-2011-0509 and CVE-2019-10799

**Comment by markusmann-vg** [comment]
Measure 1: Updated the affected dependencies
The affected dependencies should be updated to a newer version that is not affected by any known vulnerabilities.
Measure 2: Regularly check the dependencies for security issues
The dependencies should be checked regularly for known vulnerabilities. It is recommended to implement automated checks either during development or the release/deployment processes. This process should ensure that the used components are not affected by known vulnerabilities in the future.
Summary and Recommendation
It is recommended to implement all measures.

**Comment by markusmann-vg** [comment]
check also: https://github.com/hzi-braunschweig/SORMAS-Project/issues/3580

**Comment by AxelNennker** [comment]
Most dependencies are updated to their latest dependencies by this PR https://github.com/hzi-braunschweig/SORMAS-Project/pull/3626
Two are missing. I could not find newer versions.

**Comment by StefanKock** [comment]
Proposed updates:
1. com.google.guava:guava 19.0 -> 30.0-jre | 30.0-android
2. com.googlecode.libphonenumber:libphonenumber 8.12.11 -> 8.12.14
3. com.sun.xml.fastinfoset:FastInfoset 1.2.15 -> 1.2.18
4. com.vaadin:vaadin-* 8.11.1 -> 8.12.1
5. com.vladmihalcea:hibernate-types-52 2.9.13 -> 2.10.0
6. org.hibernate:hibernate-core 5.4.18.Final -> 5.4.25.Final
7. commons-beanutils:commons-beanutils 1.9.2 -> 1.9.4
8. commons-codec:commons-codec 1.9 -> 1.15
9. commons-validator:commons-validator 1.6 -> 1.7
10. io.swagger.core.v3:swagger-jaxrs2 2.1.2 -> 2.1.6
11. joda-time:joda-time 2.9.4 -> 2.10.8
12. junit:junit 4.12 -> 4.13.1
13. org.apache.commons:commons-lang3 3.10 -> 3.11
14. org.apache.httpcomponents:httpclient 4.5.4 -> 4.5.13
15. org.apache.httpcomponents:httpcore 4.4.1 -> 4.4.14
16. org.bouncycastle:* 1.66 -> 1.67
17. org.freemarker:freemarker 2.3.20 -> 2.3.30
18. org.geotools:gt-shapefile 21.4 -> 21.5
19. org.glassfish.jaxb:jaxb-runtime 2.3.1 -> 2.3.3
20. org.glassfish.jaxb:txw2 2.3.1 -> 2.3.3
21. org.hibernate.validator:hibernate-validator 6.1.5.Final -> 6.1.6.Final
22. org.jvnet.staxex:stax-ex 1.8 -> 1.8.3
23. org.keycloak:* 11.0.0 -> 11.0.3
24. org.mockito:mockito-inline 3.5.13 -> 3.5.15
25. org.postgresql:postgresql 42.2.14 -> 42.2.18
26. p6spy | p6spy | 3.9.0 -> 3.9.1

Payara modules/dependent versions:
- com.fasterxml.jackson:* 2.10.2 (bound to Payara)
- com.sun.activation:jakarta.activation 1.2.1
- com.sun.istack:istack-commons-runtime 3.0.10
- fish.payara.api:payara-api 5.2020.2
- jakarta.activation:jakarta.activation-api 1.2.1
- jakarta.xml.bind:jakarta.xml.bind-api 2.3.2

Evaluation of updates:
- ch.qos.logback:* 1.1.7 (currently on HOLD because of redeployment problems with newer versions)
- com.nexmo:client 3.3.0 (SMS library)

Is this needed/used?
- com.vaadin:vaadin-cdi 3.0.1
- com.vaadin:vaadin-spring 3.2.1
- com.vaadin:vaadin-spring-boot-starter 3.2.1
- com.vaadin:vaadin-testbench-core 5.2.0

**Comment by StefanKock** [comment]
Next step: Review/discuss list with @MartinWahnschaffeSymeda

**Comment by AxelNennker** [comment]
@StefanKock not sure whether the question is for me or somebody else
```
Is this needed/used?

    com.vaadin:vaadin-cdi 3.0.1
    com.vaadin:vaadin-spring 3.2.1
    com.vaadin:vaadin-spring-boot-starter 3.2.1
    com.vaadin:vaadin-testbench-core 5.2.0
```

I updated most of the dependencies because the owasp plugin complained about security issues.
If you uncomment that plugin in sormas-base/pom.xml and remove the mentioned updates and owasp is ok without them then the update is not needed. But I am quite sure I did only update what was needed. It might be that something I updated later already uses an updated version of something I updated earlier.
I think that vaadin8 is using really old dependencies (gwt) and sormas should update to vaadin14. Although I don't know whether vaadin14 is better in that regard.
There is a mechanism to exclude the last two insecure dependencies form the owasp plugin. So you can have owasp always check during each build whether every is up-to-date security-wise. (The vulnerabiltiy db is downloaded only once every 24h)

Axel

Why is nobody reviewing the Android PR https://github.com/hzi-braunschweig/SORMAS-Project/pull/3627 ? Don't you need Firebase? The Firebase version the pre-PR Android app is using is not working since mid November.

**Comment by StefanKock** [comment]
@AxelNennker :
- I did the list for me and to review it with @MartinWahnschaffeSymeda.
- For OWASP we are using Jenkins with decoupled update job for NVD database, it will not be included directly into the build to avoid wait time for database updates.
- The scope of this ticket does not include: Updating versions of Payara or Vaadin, updating tech stack of sormas-app (only compile dependencies mentioned in this ticket).

**Comment by StefanKock** [comment]
Waiting for @MartinWahnschaffeSymeda to discuss the dependencies to update.

**Comment by StefanKock** [comment]
Clarified with @MartinWahnschaffeSymeda how to proceed with the updates, the comment above will be kept uptodate.

**Comment by MartinWahnschaffeSymeda** [comment]
Also related #3733

**Comment by StefanKock** [comment]
Updated dependencies and merged it to development.

Outstanding testing:
- [ ] See if build job works @StefanKock 
- [ ] Test if SMS sending still works @BarnaBartha ?
- [ ] Test if the Android app works @raghupola-vg ?

**Comment by raghupola-vg** [comment]
Android app tested and works fine

**Comment by StefanKock** [comment]
Build of sormas-app does not work on my machine and @MateStrysewske, running into
```
Illegal class file: Class module-info is missing a super type. Class file version 53.
```
Reverting all version changes helped. Looking more closely on the message I found out that it was related to both jackson dependencies.
Changing them back to previous versions let my run into this:
```
Caused by: com.android.tools.r8.utils.AbortException: Error: Program type already present: androidx.activity.R$color
[artifact:mvn] 	at com.android.tools.r8.utils.Reporter.a(:21)
[artifact:mvn] 	at com.android.tools.r8.utils.Reporter.a(:7)
[artifact:mvn] 	at com.android.tools.r8.utils.t.a(:36)
[artifact:mvn] 	... 32 more

[artifact:mvn] FAILURE: Build failed with an exception.
[artifact:mvn] * What went wrong:
[artifact:mvn] Execution failed for task ':app:mergeDexRelease'.
[artifact:mvn] > A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
[artifact:mvn]    > com.android.builder.dexing.DexArchiveMergerException: Error while merging dex archives: 
[artifact:mvn]      Learn how to resolve the issue at https://developer.android.com/studio/build/dependencies#duplicate_classes.
[artifact:mvn]      Program type already present: androidx.activity.R$color
````
I downgraded guava and then it ran through.

**Comment by StefanKock** [comment]
Reviewing the dependency problems with @MartinWahnschaffeSymeda should happen.

**Comment by StefanKock** [comment]
Discussed with @MartinWahnschaffeSymeda concerning commit abc4802: Include this into a new ticket to update all dependencies of sormas-app (refers to dependencies and plugins, possibly update of gradle to 6.x will be needed).

**Comment by AxelNennker** [comment]
These vulnerable dependencies still were not updated

```
[ERROR] One or more dependencies were identified with vulnerabilities that have a CVSS score greater than or equal to '8.0': 
[ERROR] 
[ERROR] apache-jsp-8.0.9.M3.jar: CVE-2020-8022, CVE-2016-5388, CVE-2016-8735, CVE-2018-8014, CVE-2017-5648, CVE-2017-12617
[ERROR] apache-jsp-9.2.14.v20151106.jar: CVE-2017-7658, CVE-2017-7657
[ERROR] htmlunit-2.19.jar: CVE-2020-5529
[ERROR] htmlunit-core-js-2.17.jar: CVE-2020-5529
[ERROR] jetty-server-9.2.14.v20151106.jar: CVE-2017-7658, CVE-2017-7657
[ERROR] jetty-xml-9.2.14.v20151106.jar: CVE-2017-7658, CVE-2017-7657
[ERROR] tapestry-4.0.2.jar: CVE-2020-17531
[ERROR] websocket-client-9.2.13.v20150730.jar: CVE-2017-7658, CVE-2017-7657
[ERROR] 
```

Not all can be fixed because vaadin8 is too old and there are no updates for e.g. htmlunit and tapestry
Why are the ones that can be updated not updated?

Is there a documentation e.g. why using this version of htmlunit does not pose a security threat to Sormas?

**Comment by StefanKock** [comment]
> These vulnerable dependencies still were not updated
> 
> ```
> [ERROR] One or more dependencies were identified with vulnerabilities that have a CVSS score greater than or equal to '8.0': 
> [ERROR] 
> [ERROR] apache-jsp-8.0.9.M3.jar: CVE-2020-8022, CVE-2016-5388, CVE-2016-8735, CVE-2018-8014, CVE-2017-5648, CVE-2017-12617
> [ERROR] apache-jsp-9.2.14.v20151106.jar: CVE-2017-7658, CVE-2017-7657
> [ERROR] htmlunit-2.19.jar: CVE-2020-5529
> [ERROR] htmlunit-core-js-2.17.jar: CVE-2020-5529
> [ERROR] jetty-server-9.2.14.v20151106.jar: CVE-2017-7658, CVE-2017-7657
> [ERROR] jetty-xml-9.2.14.v20151106.jar: CVE-2017-7658, CVE-2017-7657
> [ERROR] tapestry-4.0.2.jar: CVE-2020-17531
> [ERROR] websocket-client-9.2.13.v20150730.jar: CVE-2017-7658, CVE-2017-7657
> [ERROR] 
> ```

Non of these libraries are knowingly dependencies of running code, but might be listed in `vaadin-bom`.

**Comment by AxelNennker** [comment]
Those dependencies are reported as vulnerable by the owasp plugin mentioned here
https://github.com/hzi-braunschweig/SORMAS-Project/pull/3626

If you run `mvn dependency:tree -Dverbose > /tmp/x` and search for the vulnerable packages you'll find the package that depends on them. Then go upwards in the tree and update the dependency that is mentioned on some pom.xml. If there is no match in a project's pom.xml then update the hightest package in the tree that is on the same branch as the vulnerable on. This update strategy often fixes several lower vulnerable packages.

That a package is not mentioned in the projects' pom.xml files does not mean that a vulnerable version is not used.

`mvn versions:display-dependency-updates` shows all the dependencies that can be updated.
You probably know that... I would run this command each Tuesday and consider updating the minor stuff on that day and consider updating bigger updates and create an issue for each.

**Comment by JonasCir** [comment]
@AxelNennker Thanks for raising the concern!
I currently work on a security CI for this project and I think that checking for vulnerable dependencies should be a part of it. I try to understand the issue at hand better. Right now you are concerned about vulnerable,  _transitive_ dependencies of packages that we depend on, correct?

**Comment by AxelNennker** [comment]
> @AxelNennker Thanks for raising the concern!
> I currently work on a security CI for this project and I think that checking for vulnerable dependencies should be a part of it. I try to understand the issue at hand better. Right now you are concerned about vulnerable, _transitive_ dependencies of packages that we depend on, correct?
> 
> Can confirm `mvn dependency-check:check` has findings for direct dependencies of the project. `dependency-check:aggregate` which also goes through transitive dependencies fails with a `NullPointerException`.

I also see the NPE in `dependency-check:aggregate`. But that is another issue.
Filed it here https://github.com/jeremylong/DependencyCheck/issues/3022
I never used aggregate before.

I suggest to add dependency-check:check to Sormas's CI and run it on every pull request.
For my work projects I create a scheduled CI run that checks for vulnerabilities each morning.

For Sormas I tried to fix all vulnerabilities in this PR https://github.com/hzi-braunschweig/SORMAS-Project/pull/3626
Two dependencies could not be updated to secure versions. Vaadin8 is just too old.

I suggest to update all vulnerable dependencies and to analyze the threats and suppress the remaining dependencies if they are no threat using
```
<suppressionFiles>
  <suppressionFile>project-suppression.xml</suppressionFile>
</suppressionFiles>
```

**Comment by StefanKock** [comment]
@AxelNennker @JonasCir This ticket is closed and what you are discussing here is out of scope, please stop chatting here.
Please have a look on #3580 concerning suppressions we already consider with Dependency-Check in Jenkins. If you disagree on that ticket that other not yet found dependencies are outdated please create a new ticket.

**Comment by AxelNennker** [comment]
Created https://github.com/hzi-braunschweig/SORMAS-Project/issues/3822 as requested by @StefanKock
